lib/ansible/module_utils/network/nxos/facts/telemetry/telemetry.py

```diff
@@ -0,0 +1,164 @@
+        device_cache = ref.cache_existing
+
+        if device_cache is None:
+            device_cache_lines = []
+        else:
+            device_cache_lines = device_cache.split("\n")
+
+        # Get Telemetry Destination Group Data
+        cmd_ref['TMS_DESTGROUP']['ref'] = []
+        for line in device_cache_lines:
+            if re.search(r'destination-group', line):
+                resource_key = line.strip()
+                cmd_ref['TMS_DESTGROUP']['ref'].append(NxosCmdRef(self._module, TMS_DESTGROUP))
+                ref = cmd_ref['TMS_DESTGROUP']['ref'][-1]
+                ref.set_context([resource_key])
+                ref.get_existing(device_cache)
+                normalize_data(ref)
+
+        # Get Telemetry Sensorgroup Group Data
+        cmd_ref['TMS_SENSORGROUP']['ref'] = []
+        for line in device_cache_lines:
+            if re.search(r'sensor-group', line):
+                resource_key = line.strip()
+                cmd_ref['TMS_SENSORGROUP']['ref'].append(NxosCmdRef(self._module, TMS_SENSORGROUP))
+                ref = cmd_ref['TMS_SENSORGROUP']['ref'][-1]
+                ref.set_context([resource_key])
+                ref.get_existing(device_cache)
+
+        # Get Telemetry Subscription Data
+        cmd_ref['TMS_SUBSCRIPTION']['ref'] = []
+        for line in device_cache_lines:
+            if re.search(r'subscription', line):
+                resource_key = line.strip()
+                cmd_ref['TMS_SUBSCRIPTION']['ref'].append(NxosCmdRef(self._module, TMS_SUBSCRIPTION))
+                ref = cmd_ref['TMS_SUBSCRIPTION']['ref'][-1]
+                ref.set_context([resource_key])
+                ref.get_existing(device_cache)
+
+        objs = []
+        objs = self.render_config(self.generated_spec, cmd_ref)
+        facts = {'telemetry': {}}
+        if objs:
+            # params = utils.validate_config(self.argument_spec, {'config': objs})
+            facts['telemetry'] = objs
+
+        ansible_facts['ansible_network_resources'].update(facts)
+        if self._module.params.get('state'):
+            self._module.params['state'] = saved_module_state
+        return ansible_facts
+
+    def render_config(self, spec, cmd_ref):
+        """
+        Render config as dictionary structure and delete keys
+          from spec for null values
+
+        :param spec: The facts tree, generated from the argspec
+        :param conf: The configuration
+        :rtype: dictionary
+        :returns: The generated config
+        """
+        config = deepcopy(spec)
+        config['destination_groups'] = []
+        config['sensor_groups'] = []
+        config['subscriptions'] = []
+        managed_objects = ['TMS_GLOBAL', 'TMS_DESTGROUP', 'TMS_SENSORGROUP', 'TMS_SUBSCRIPTION']
+
+        # Walk the argspec and cmd_ref objects and build out config dict.
+        for key in config.keys():
+            cr_keys = [key]
+            for mo in managed_objects:
+                for cr in cmd_ref[mo]['ref']:
```

Comment:
I'm unsure about the usage here... isn't `'ref'` a dict? 
Will there ever be more than one `'ref'` for a given `mo`? 
Or is this for multiple contexts?

Reply:
No `'ref'` is a list of `CmdRef` objects.  So yes, it's for multiple contexts. For example,  in the code above, I create a `CmdRef` object for each sensorgroup I find and add it to a list.

In this case it would be:

`cmd_ref['TMS_SENSORGROUP']['ref']` where this is a list of sensorgroup `CmdRef` objects.

Reply:
👍